### PR TITLE
Aggregate journals in APIv3

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## OpenProject work package

https://community.openproject.org/work_packages/20688
## Description

Starting with this PR the APIv3 activities return their results in an aggregated form.
Some work on the `Journal::AggregatedJournals` was still neccessary.

**WP legcy form (non-aggregated)**
non-aggregated

**inplace view (aggregated)**
aggregated

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-aggregated-api-journals.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-aggregated-api-journals.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by oliverguenther** [comment]
I've restarted the specs thrice in the course over late Friday / Saturday as they interfered with my spec run. The Postgres timeout is persistent. Some change seems to successfully lock up the DB.

**Comment by NobodysNightmare** [comment]
PostgreSQL is weird... I already encountered this earlier: When you have multiple specs with a bad SQL syntax it will (after ~5 bad queries) start to refuse further connections, letting rails time out on each and every test case.

For travis this means hitting the 120 minute mark, never telling anything about the connection timeouts.

**Comment by oliverguenther** [comment]
Travis did tell me about a timeout on Postgres, something like `couldn't get stable pg connection for 5 minutes ..`, though these messages seem to disappear.

File: app/models/journal/aggregated_journal.rb [hunk 1]
```diff
@@ -182,21 +202,31 @@ def table_name
...
   end
 
   def predecessor
     unless defined? @predecessor
+      version_sql = "COALESCE(addition.version, #{Journal.table_name}.version)"
       raw_journal = self.class.query_aggregated_journals(journable: journable)
-                    .where("#{Journal.table_name}.version < ?", version)
-                    .order("#{Journal.table_name}.version DESC")
+                    .where("#{version_sql} < ?", version)
+                    .except(:order)
```
Comment (Azure7111) [inline app/models/journal/aggregated_journal.rb:228]: Do I get this right: This line removes, or rather prevents, an ordering which happens inside `query_aggregated_journals`?

Comment (NobodysNightmare) [inline app/models/journal/aggregated_journal.rb:228]: Right. I had to look this up myself:

When you `order`, then AR will only **append** that order to the existing order:
`order(id: :asc).order(name: :desc) == 'ORDER BY id ASC, name DESC'`

By using `except(:order)` you throw away anything there was about the order and start from scratch.

The old (deprecated?) way of doing it is `reorder`

Comment (Azure7111) [inline app/models/journal/aggregated_journal.rb:228]: Given the fact that orders are _not_ "overwritten" this makes actually sense to me now. Thanks!

File: lib/api/v3/activities/activities_by_work_package_api.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  module V3
+    module Activities
+      class ActivitiesByWorkPackageAPI < ::API::OpenProjectAPI
+        resource :activities do
+          helpers do
+            def save_work_package(work_package)
+              if work_package.save
+                journals = ::Journal::AggregatedJournal.aggregated_journals(
+                  journable: work_package)
+                Activities::ActivityRepresenter.new(journals.last, current_user: current_user)
```
Comment (Azure7111) [inline lib/api/v3/activities/activities_by_work_package_api.rb:41]: Can we be sure that no other journals have been created since the last line?

File: lib/api/v3/activities/activity_representer.rb [hunk 3]
```diff
@@ -47,20 +48,20 @@ class ActivityRepresenter < ::API::Decorators::Single
...
-            title: "#{represented.user.name} - #{represented.user.login}"
+            href: api_v3_paths.user(represented.user.id)
           }
         end
 
         link :update do
           {
-            href: api_v3_paths.activity(represented.id),
-            method: :patch,
-            title: "#{represented.id}"
```
Comment (Azure7111) [inline lib/api/v3/activities/activity_representer.rb:59]: Why did you remove the link titles?

File: lib/plugins/acts_as_journalized/lib/journal_changes.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+#-- encoding: UTF-8
+module JournalChanges
+  def get_changes
+    return {} if data.nil?
```
Comment (Azure7111) [inline lib/plugins/acts_as_journalized/lib/journal_changes.rb]: I'm not sure about this: Is there any way to ensure that the including class of this module has a `data` method?

File: lib/plugins/acts_as_journalized/lib/journal_changes.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    return @changes if @changes
+
+    @changes = HashWithIndifferentAccess.new
+
+    if predecessor.nil?
+      @changes = data.journaled_attributes.select { |_, v| !v.nil? }
+                   .inject({}) { |h, (k, v)| h[k] = [nil, v]; h }
```
Comment (Azure7111) [inline lib/plugins/acts_as_journalized/lib/journal_changes.rb]: I have to agree, this needs more style. Please try again.

File: lib/api/v3/activities/activities_by_work_package_api.rb (hunk 2 above)
Comment (NobodysNightmare) [inline lib/api/v3/activities/activities_by_work_package_api.rb:41]: I don't think so. But:
- I think there is no quick way to do it better
- I did not introduce this problem, it was there before

This does not mean it shouldn't be solved, it just does not belong here and I hope that it is for now unimportant enough.

File: lib/api/v3/activities/activity_representer.rb (hunk 3 above)
Comment (NobodysNightmare) [inline lib/api/v3/activities/activity_representer.rb:59]: Because `16435` is not a title I would present to a user.

And I wont introduce a proper shiny formatting (including localization) for a title that no one (assumption) will ever display.

File: lib/plugins/acts_as_journalized/lib/journal_changes.rb (hunk 4 above)
Comment (NobodysNightmare) [inline lib/plugins/acts_as_journalized/lib/journal_changes.rb]: In case it does not have one, a `NoMethodError` will be thrown to indicate that you did it wrong xD

I guess the answer here is "it is a dynamic world, deal with it".
I've seen annotations and stuff in other environments, but that requires your tools to support you.

The other option one could choose is to pass all needed values as parameter into `get_changes`. However, this is an approach I would use when designing a class that the caller will instantiate and hold as member (composition). In this case I chose to use a module that another class can include (which is more Trait-like). In that case I prefer to go the "expect what you need to be there" route.

File: lib/plugins/acts_as_journalized/lib/journal_changes.rb (hunk 5 above)
Comment (NobodysNightmare) [inline lib/plugins/acts_as_journalized/lib/journal_changes.rb]: Not my darn code, I solely moved it.

File: spec/requests/api/v3/activity_resource_spec.rb [hunk 6]
```diff
@@ -80,9 +84,40 @@
...
+    it_behaves_like 'handling anonymous user' do
       let(:project) { FactoryGirl.create(:project, is_public: true) }
-      let(:id) { activity.id }
+      let(:path) { api_v3_paths.activity activity.id }
+    end
+  end
+
+  describe '#patch' do
+    subject(:response) { last_response }
+    let(:patch_path) { api_v3_paths.activity activity.id }
```
Comment (Azure7111) [inline spec/requests/api/v3/activity_resource_spec.rb:95]: Do you think the patch_path and the get_path are likely to diverge in the near future? Because if not, then you should be able to move it into the initialization section in the beginning of this file. But if you say that this is about consistency across the recource specs I am fine with that too.

File: spec/requests/api/v3/support/authorization.rb [hunk 7]
```diff
@@ -28,27 +28,16 @@
...
-      before { get get_path }
-
-      it 'should respond with 200' do
-        expect(subject.status).to eq(200)
-      end
-
-      it 'should respond with correct type' do
-        expect(subject.body).to include_json(type.to_json).at_path('_type')
-        expect(subject.body).to be_json_eql(id.to_json).at_path('id')
-      end
```
Comment (Azure7111) [inline spec/requests/api/v3/support/authorization.rb:45]: Why have you removed these tests?

File: spec/requests/api/v3/activity_resource_spec.rb (hunk 6 above)
Comment (NobodysNightmare) [inline spec/requests/api/v3/activity_resource_spec.rb:95]: I actually thought that it is how all specs do it and that is why I sticked to it.

Personally I like the generic `path` more. You won't issue two different requests in one spec anyway.

File: spec/requests/api/v3/support/authorization.rb (hunk 7 above)
Comment (NobodysNightmare) [inline spec/requests/api/v3/support/authorization.rb:45]: Because they are bogus.
They are not connected to the anonymous user in any way, but to "a user with sufficient permissions", whcih is always tested in the main spec.

What is not in here (and usually in the main spec) is the "user with insufficient permissions". Note that I don't use the word "anonymous user" here, because it just has nothing to do with being anonymous.

However, there is one setting that is related to being anonymous, regardless of other permissions: Receiving a 401 as anonymous user, when the setting `login_required?` is enabled. That is why I left the one remaining spec alive, because I could see its value.

The other specs only duplicate the stuff of the main specs while confusing people by using the term "anonymous user".

**Comment by Azure7111** [comment]
The requirements talk about "the angular client (inplace edit) is adapted accordingly". I guess that means being able to see aggregated journals in the details pane, or is there more to it?

Otherwise this LGTM :+1:

**Comment by NobodysNightmare** [comment]
> I guess that means being able to see aggregated journals in the details pane

Right, but I did not want to write "the angular client is not broken afterwards", which is what I basically wanted to say ^^

**Comment by Azure7111** [comment]
Alright, then I will merge this once it's green.
